prost-derive: Add #[(prost(prost_path = "::prost")]

jeremija opened the pull request:
**Work in progress**

Add a new proc_macro attribute in `prost-derive`, which is used to let the macro implementation know about the value of
`prost_build::Config::prost_build`.

As a result, the generated types no longer have `::prost` hardcoded in them, and a re-exported prost crate can be used instead along with generated code.

Fixes #939

jeremija commented:
I didn't realize there was already #488 which basically does the same thing, and adds additional fix for `prost_types`.
I'd really want to get a fix for the original issue merged and fixed, so please let me know if there's anything I can do to help.

chris-zen commented:
Do you have any plans to move this or #488 forward? I'm looking to wrap prost in my library so users don’t need to worry about the correct version—they can just rely on my library instead.

jeremija commented:
I've updated this MR to also include the ability to re-export `prost_types` if desired. Also added a separate test crate to ensure the re-exports work as intended.

demurgos commented:
I have an issue at work where my binary crate depends on two libs working with generated Prost code, except they use different major versions of Prost. With the hardcoded `prost` path, I can't fix it and have to wait until they update to the same version. However, any time a new Prost version is released and the libs get out of sync I'm back to being stuck.

With a way to override the path, I could use cargo to depend on both dependencies:
```
prost014 = {package="prost", version "0.14"}
prost015 = {package="prost", version "0.15"}
```

And it would avoid global coordination issues.

This PR would solve this issue neatly. :+1:

caspermeijn commented on tests/prost-path/src/lib.rs:
```diff
@@ -0,0 +1 @@
+include!(concat!(env!("OUT_DIR"), "/prost_path.rs"));
```
This test doesn't actually test anything. Please verify the correct types are used and explain what this test is testing.

jeremija replied:
The dependencies in `Cargo.toml` only include the `reexported_prost` crate, and the configuration in `build.rs` contains the following:

```rust
fn main() {
    prost_build::Config::new()
        .prost_path("::reexported_prost::prost")
        .prost_types_path("::reexported_prost::prost_types")
        .compile_protos(&["protos/prost_path.proto"], &["protos"])
        .unwrap();
}
```

Therefore if it successfully compiles, it means that the change is working as expected.

Is there something I'm missing here, or would it be enough to just add that as a comment above?

jeremija commented on prost-derive/src/lib.rs:
```diff
@@ -519,6 +521,64 @@ pub fn oneof(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+                }
+                _ => continue,
+            }
+        }
```
This should be done now, let me know what you think!

jeremija commented:
@caspermeijn did you get a chance to have a look at this? would be great to have this merged and I'm eager to address any remaining concerns you might have. TIA

caspermeijn commented on prost-build/src/context.rs:
```diff
@@ -43,7 +50,11 @@ impl<'a> Context<'a> {
 
     pub fn prost_path(&self) -> &str {
-        self.config.prost_path.as_deref().unwrap_or("::prost")
+        self.config.prost_path_or_default()
```
I see, the field `config` is not public, so the helper function is needed. I was about to suggest removing one of the two and just call the other one. But I see the is difficult.

caspermeijn reviewed (approved):
Thank you for your contribution
